### PR TITLE
[SPARK-40759][SQL] Migrate type check failures of time window onto error classes

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr replace `TypeCheckFailure` by `DataTypeMismatch` in type checks in the time window expressions:

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing error messages.

### How was this patch tested?

- Pass GitHub Actions
- Manual test:

```
build/sbt "catalyst/testOnly org.apache.spark.sql.catalyst.analysis.AnalysisErrorSuite" -Pscala-2.13
```
passed

### CODE REVIEW

**Comment by LuciferYang** [comment]
Still WIP, will fix test fail later

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -270,6 +270,11 @@
           "The <exprName> must not be null"
         ]
       },
+      "UNEXPECTED_PARAMETER_CONSTRAINT" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:273]: How about `PARAMETER_CONSTRAINT_VIOLATION`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala [hunk 2]
```diff
@@ -114,18 +115,48 @@ case class TimeWindow(
     val dataTypeCheck = super.checkInputDataTypes()
     if (dataTypeCheck.isSuccess) {
       if (windowDuration <= 0) {
-        return TypeCheckFailure(s"The window duration ($windowDuration) must be greater than 0.")
+        return DataTypeMismatch(
+          errorSubClass = "VALUE_OUT_OF_RANGE",
+          messageParameters = Map(
+            "exprName" -> "window_duration",
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:121]: Please, use `toSQLId()` here and below.

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (LuciferYang) [reply core/src/main/resources/error/error-classes.json:273]: fine to me

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala (hunk 2 above)
Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:121]: OK

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala [hunk 3]
```diff
@@ -433,67 +433,131 @@ class AnalysisErrorSuite extends AnalysisTest {
...
   )
 
-  errorTest(
+  errorClassTest(
     "negative window duration in time window",
     testRelation.select(
       TimeWindow(Literal("2016-01-01 01:01:01"), "-1 second", "1 second", "0 second").as("window")),
-      "The window duration " :: " must be greater than 0." :: Nil
+      "DATATYPE_MISMATCH.VALUE_OUT_OF_RANGE",
+    Map(
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:516]: Discuss: @MaxGekk 

There is a problem here: if the order of the map content changes from

```scala
    Map(
      "exprName" -> "`slide_duration`",
      "valueRange" -> "(0, 9223372036854775807]",
      "currentValue" -> "-1000000L",
      "sqlExpr" -> "\"window(2016-01-01 01:01:01, 1000000, -1000000, 0)\""
    )
```


Changed to

```scala
    Map(
      "sqlExpr" -> "\"window(2016-01-01 01:01:01, 1000000, -1000000, 0)\"",
      "exprName" -> "`slide_duration`",
      "valueRange" -> "(0, 9223372036854775807]",
      "currentValue" -> "-1000000L" 
    )
```

the assertion will fail, but is the order important? seems too strict, WDYT?

If the order is not so important, I want to refactor the relevant check logic

Comment (LuciferYang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:516]: But if the ` map If size>4 ', the result of `sameElements` is independent of the order:

For example:

```scala
val map1 = Map(
      "exprName" -> "`window_duration`",
      "valueRange" -> s"(0, 9223372036854775807]",
      "currentValue" -> "-1000000L",
      "sqlExpr" -> "\"window(2016-01-01 01:01:01, -1000000, 1000000, 0)\""
    )
val map2 =  Map(
      "sqlExpr" -> "\"window(2016-01-01 01:01:01, -1000000, 1000000, 0)\"",
      "exprName" -> "`window_duration`",
      "valueRange" -> s"(0, 9223372036854775807]",
      "currentValue" -> "-1000000L"
    )

map1.sameElements(map2)
```
the result is `false`.

but

```scala
val map1 = Map(
      "exprName" -> "`window_duration`",
      "valueRange" -> s"(0, 9223372036854775807]",
      "currentValue" -> "-1000000L",
      "sqlExpr" -> "\"window(2016-01-01 01:01:01, -1000000, 1000000, 0)\"",
       "1" -> "1"
    )
val map2 =  Map(
      "sqlExpr" -> "\"window(2016-01-01 01:01:01, -1000000, 1000000, 0)\"",
      "exprName" -> "`window_duration`",
      "valueRange" -> s"(0, 9223372036854775807]",
      "currentValue" -> "-1000000L",
      "1" -> "1"
    )

map1.sameElements(map2)
```
the result is `true`.

I think we should refactor the check logic, I will do it later:

https://github.com/apache/spark/blob/bef979757bd12ecbe706aa14ec7e09cd84d03d76/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala#L197-L202

Comment (LuciferYang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:516]: try fix by https://github.com/apache/spark/pull/38396

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala [hunk 4]
```diff
@@ -433,67 +433,131 @@ class AnalysisErrorSuite extends AnalysisTest {
...
 
-  errorTest(
+  errorClassTest(
     "zero slide duration in time window",
     testRelation.select(
       TimeWindow(Literal("2016-01-01 01:01:01"), "1 second", "0 second", "0 second").as("window")),
-      "The slide duration" :: " must be greater than 0." :: Nil
+    "DATATYPE_MISMATCH.VALUE_OUT_OF_RANGE",
+    Map(
+      "sqlExpr" -> "\"window(2016-01-01 01:01:01, 1000000, 0, 0)\"",
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:556]: rebase and move `sqlExpr` to the first

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala [hunk 5]
```diff
@@ -114,18 +115,48 @@ case class TimeWindow(
...
+          )
+        )
       }
       if (startTime.abs >= slideDuration) {
-        return TypeCheckFailure(s"The absolute value of start time ($startTime) must be less " +
-          s"than the slideDuration ($slideDuration).")
+        return DataTypeMismatch(
+          errorSubClass = "PARAMETER_CONSTRAINT_VIOLATION",
+          messageParameters = Map(
+            "leftExprName" -> toSQLId("start_time"),
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:153]: Precisely speaking, it should be `abs(start_time)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala [hunk 6]
```diff
@@ -114,18 +115,48 @@ case class TimeWindow(
...
+            "leftExprValue" -> toSQLValue(startTime, LongType),
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:154]: Actually, `startTime.abs`

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @LuciferYang.

**Pull request closed without merging**